File: internal/orchestration/router_test.go
package orchestration

import (
	"testing"
)

func TestSimilarity(t *testing.T) {
	if similarity("research", "research") != 1.0 {
		t.Error("identical strings should return 1.0")
	}
	if similarity("research", "xyz") >= 0.3 {
		t.Error("unrelated strings should score < 0.3")
	}
	if similarity("prospecting", "prospect") < 0.5 {
		t.Error("similar strings should score > 0.5")
	}
}

func TestTaskRouter_RegisterUnregister(t *testing.T) {
	r := NewTaskRouter()
	r.RegisterAgent("alice", []SkillDeclaration{{Name: "research", Proficiency: 0.9}})
	if _, ok := r.agents["alice"]; !ok {
		t.Fatal("agent should be registered")
	}
	r.UnregisterAgent("alice")
	if _, ok := r.agents["alice"]; ok {
		t.Fatal("agent should be removed")
	}
}

func TestTaskRouter_ScoreMatch(t *testing.T) {
	r := NewTaskRouter()
	r.RegisterAgent("alice", []SkillDeclaration{
		{Name: "market-research", Proficiency: 0.9},
		{Name: "competitive-analysis", Proficiency: 0.8},
	})

	task := TaskDefinition{
		ID:             "t1",
		RequiredSkills: []string{"market-research"},
	}
	score := r.ScoreMatch("alice", task)
	if score < 0.3 {
		t.Errorf("expected score >= 0.3, got %f", score)
	}

	// Unknown agent returns 0.
	if r.ScoreMatch("unknown", task) != 0 {
		t.Error("unknown agent should return 0")
	}

	// Empty required skills returns 0.
	emptyTask := TaskDefinition{ID: "t2"}
	if r.ScoreMatch("alice", emptyTask) != 0 {
		t.Error("empty required skills should return 0")
	}
}

func TestTaskRouter_FindBestAgent(t *testing.T) {
	r := NewTaskRouter()
	r.RegisterAgent("alice", []SkillDeclaration{{Name: "prospecting", Proficiency: 0.9}})
	r.RegisterAgent("bob", []SkillDeclaration{{Name: "outreach", Proficiency: 0.7}})

	task := TaskDefinition{
		ID:             "t1",
		RequiredSkills: []string{"prospecting"},
	}
	best := r.FindBestAgent(task)
	if best == nil {
		t.Fatal("expected a best agent")
	}
	if best.AgentSlug != "alice" {
		t.Errorf("expected alice, got %s", best.AgentSlug)
	}

	// No match returns nil.
	noMatchTask := TaskDefinition{
		ID:             "t2",
		RequiredSkills: []string{"zzz-nonexistent"},
	}
	if r.FindBestAgent(noMatchTask) != nil {
		t.Error("expected nil for unmatched task")
	}
}

func TestTaskRouter_FindCapableAgents_Sorted(t *testing.T) {
	r := NewTaskRouter()
	r.RegisterAgent("alice", []SkillDeclaration{{Name: "seo", Proficiency: 0.5}})
	r.RegisterAgent("bob", []SkillDeclaration{{Name: "seo", Proficiency: 1.0}})

	task := TaskDefinition{
		ID:             "t1",
		RequiredSkills: []string{"seo"},
	}
	results := r.FindCapableAgents(task)
	if len(results) != 2 {
		t.Fatalf("expected 2 capable agents, got %d", len(results))
	}
	// Should be sorted descending — bob has higher proficiency.
	if results[0].Score < results[1].Score {
		t.Error("results should be sorted descending by score")
	}
}

File: cli/tests/orchestration/router.test.ts
import { describe, it, beforeEach } from 'node:test';
import assert from 'node:assert/strict';
import { TaskRouter } from '../../src/orchestration/router.js';
import type { TaskDefinition } from '../../src/orchestration/types.js';

function makeTask(overrides: Partial<TaskDefinition> = {}): TaskDefinition {
  return {
    id: 'task-1',
    title: 'Test task',
    description: 'A test task',
    requiredSkills: [],
    priority: 'medium',
    status: 'pending',
    createdAt: Date.now(),
    ...overrides,
  };
}

describe('TaskRouter', () => {
  let router: TaskRouter;

  beforeEach(() => {
    router = new TaskRouter();
  });

  // ── findBestAgent ─────────────────────────────────────────────────

  it('findBestAgent returns agent with highest skill match', () => {
    router.registerAgent('seo-bot', [
      { name: 'seo', description: 'SEO', proficiency: 0.9 },
      { name: 'keyword-research', description: 'Keywords', proficiency: 0.8 },
    ]);
    router.registerAgent('sales-bot', [
      { name: 'prospecting', description: 'Sales', proficiency: 0.9 },
    ]);

    const task = makeTask({ requiredSkills: ['seo', 'keyword-research'] });
    const result = router.findBestAgent(task);
    assert.ok(result);
    assert.equal(result.agentSlug, 'seo-bot');
  });

  it('findBestAgent returns null when no agents registered', () => {
    const task = makeTask({ requiredSkills: ['seo'] });
    const result = router.findBestAgent(task);
    assert.equal(result, null);
  });

  it('findBestAgent returns null when no agents match skills', () => {
    router.registerAgent('sales-bot', [
      { name: 'prospecting', description: 'Sales', proficiency: 0.9 },
    ]);

    const task = makeTask({ requiredSkills: ['quantum-physics'] });
    const result = router.findBestAgent(task);
    assert.equal(result, null);
  });

  // ── findCapableAgents ─────────────────────────────────────────────

  it('findCapableAgents returns sorted list', () => {
    router.registerAgent('seo-bot', [
      { name: 'seo', description: 'SEO', proficiency: 0.9 },
    ]);
    router.registerAgent('content-bot', [
      { name: 'seo', description: 'SEO basics', proficiency: 0.5 },
    ]);
    router.registerAgent('sales-bot', [
      { name: 'prospecting', description: 'Sales', proficiency: 0.9 },
    ]);

    const task = makeTask({ requiredSkills: ['seo'] });
    const results = router.findCapableAgents(task);

    assert.ok(results.length >= 2);
    // First should have higher score than second
    assert.ok(results[0].score >= results[1].score);
    assert.equal(results[0].agentSlug, 'seo-bot');
  });

  it('findCapableAgents excludes agents with 0 matching skills', () => {
    router.registerAgent('seo-bot', [
      { name: 'seo', description: 'SEO', proficiency: 0.9 },
    ]);
    router.registerAgent('sales-bot', [
      { name: 'prospecting', description: 'Sales', proficiency: 0.9 },
    ]);

    const task = makeTask({ requiredSkills: ['seo'] });
    const results = router.findCapableAgents(task);

    // sales-bot should not appear (score 0)
    const slugs = results.map((r) => r.agentSlug);
    assert.ok(!slugs.includes('sales-bot'));
  });

  // ── scoreMatch ────────────────────────────────────────────────────

  it('agent with 0 matching skills gets score 0', () => {
    router.registerAgent('sales-bot', [
      { name: 'prospecting', description: 'Sales', proficiency: 0.9 },
    ]);

    const task = makeTask({ requiredSkills: ['quantum-physics'] });
    const score = router.scoreMatch('sales-bot', task);
    assert.equal(score, 0);
  });

  it('task with no required skills matches any agent with score 1', () => {
    router.registerAgent('generic-bot', [
      { name: 'general', description: 'General', proficiency: 0.5 },
    ]);

    const task = makeTask({ requiredSkills: [] });
    const score = router.scoreMatch('generic-bot', task);
    assert.equal(score, 1);
  });

  it('scoreMatch returns 0 for unregistered agent', () => {
    const task = makeTask({ requiredSkills: ['seo'] });
    const score = router.scoreMatch('nonexistent', task);
    assert.equal(score, 0);
  });

  it('higher proficiency yields higher score', () => {
    router.registerAgent('expert', [
      { name: 'seo', description: 'SEO', proficiency: 0.9 },
    ]);
    router.registerAgent('novice', [
      { name: 'seo', description: 'SEO', proficiency: 0.3 },
    ]);

    const task = makeTask({ requiredSkills: ['seo'] });
    const expertScore = router.scoreMatch('expert', task);
    const noviceScore = router.scoreMatch('novice', task);
    assert.ok(expertScore > noviceScore);
  });

  // ── unregisterAgent ───────────────────────────────────────────────

  it('unregisterAgent removes the agent', () => {
    router.registerAgent('seo-bot', [
      { name: 'seo', description: 'SEO', proficiency: 0.9 },
    ]);
    router.unregisterAgent('seo-bot');

    const task = makeTask({ requiredSkills: ['seo'] });
    const result = router.findBestAgent(task);
    assert.equal(result, null);
  });

  // ── fuzzy matching ────────────────────────────────────────────────

  it('fuzzy matches similar skill names', () => {
    router.registerAgent('enricher', [
      { name: 'data-enrichment', description: 'Enrichment', proficiency: 0.8 },
    ]);

    // "enrichment" should fuzzy-match "data-enrichment"
    const task = makeTask({ requiredSkills: ['enrichment'] });
    const score = router.scoreMatch('enricher', task);
    assert.ok(score > 0, `Expected score > 0, got ${score}`);
  });

  it('agent with zero skills gets score 0 for any task with requirements', () => {
    router.registerAgent('empty-bot', []);
    const task = makeTask({ requiredSkills: ['seo'] });
    const score = router.scoreMatch('empty-bot', task);
    assert.equal(score, 0);
  });
});
